### ISSUE TITLE
Compiler failure

### ISSUE THREAD

**Issue by dpomier** [HaxeFoundation/haxe#10261]
```haxe
function main() {
	read();
}

inline function read() {
	while (true) Sys.sleep(1);
	return haxe.io.Bytes.alloc(0);
}

function write() {
	return read();
}
```
Results in:
```
Compiler failure
Please submit an issue at https://github.com/HaxeFoundation/haxe/issues/new
Attach the following information:
Haxe: 4.2.2; OS type: unix;
File "src/optimization/analyzerTexprTransformer.ml", line 326, characters 17-24
Called from file "src/optimization/analyzerTexprTransformer.ml", line 184, characters 13-24
Called from file "src/optimization/analyzerTexprTransformer.ml", line 537, characters 16-27
Called from file "src/optimization/analyzerTexprTransformer.ml", line 535, characters 4-39
Called from file "src/optimization/analyzerTexprTransformer.ml", line 631, characters 14-32
Called from file "src/optimization/analyzerTexprTransformer.ml", line 659, characters 15-39
Called from file "src/optimization/analyzerTexprTransformer.ml", line 667, characters 23-47
Called from file "src/optimization/analyzer.ml", line 922, characters 10-13
Called from file "src/optimization/analyzer.ml", line 961, characters 2-126
Called from file "src/optimization/analyzer.ml", line 1025, characters 25-37
Called from file "src/optimization/analyzer.ml", line 1057, characters 4-22
Called from file "list.ml", line 85, characters 12-15
Called from file "src/optimization/analyzer.ml", line 1081, characters 2-53
Called from file "list.ml", line 85, characters 12-15
Called from file "src/optimization/analyzer.ml", line 922, characters 10-13
Called from file "src/filters/filters.ml", line 783, characters 31-71
Called from file "src/compiler/haxe.ml", line 611, characters 1-26
Called from file "src/compiler/haxe.ml", line 1078, characters 2-39
Called from file "src/compiler/haxe.ml", line 672, characters 3-11
Called from file "src/compiler/haxe.ml", line 1247, characters 1-35
File "src/core/globals.ml", line 117, characters 1-7: Assertion failed
```
https://try.haxe.org/#6BB04EBa

There are a few changes that seem to walk around the assertion failure:
- removing `inline`,
- removing the function declaration `write`,
- using `Int` instead of `haxe.io.Bytes`,
- replacing `while(true)` for a `if` or `while(false)`, etc.

**Comment by Simn** [comment]
Could you post the compilation parameters you're using? I can't seem to reproduce this.

**Comment by dpomier** [comment]
This should be enough: `-main Main --interp`. It seems js works, while eval/hl/jvm/etc don't.

**Comment by Simn** [comment]
Ok your two cases confused me. Case 1 is just a normal infinite loop, it hangs with `--interp` as expected because what else would it do.

But yes, I can reproduce this in case 2. Took me a minute to notice the additional `return` in that `write` function, which makes the difference here.

**Comment by dpomier** [comment]
I see, the `compilation failed` error on try.haxe.org confused me for Case 1. As it works as expected locally I removed it from the description.

**Comment by Simn** [comment]
I've managed to remove some distractions:

```haxe
function main() {

}

function call(x:Int) {
	return "";
}

function write() {
	return {
		while (true) {
		}
		call(0);
	}
}
```

This is some wacky code where we expect a value but within that value come across unreachable code, and then the compiler gets really confused. IMO this _should_ be a compilation error, but the last time I tried to give errors from dead code everyone yelled at me.

Given how silly this example looks, I wonder if you oversimplified it from your actual issue. If not, you seem to have some dead code after an infinite loop.
